shared/src/lib.rs

```diff
@@ -50,6 +51,21 @@ extern crate shrinkwraprs;
 #[macro_use]
 extern crate derive_new;
 
+type PravegaNodeUriParseResult<T> = std::result::Result<T, PravegaNodeUriParseError>;
+
+#[derive(Debug, Snafu)]
+pub enum PravegaNodeUriParseError {
+    #[snafu(display("Could not parse uri due to {}", error_msg))]
+    ParseError { error_msg: String },
+}
+
+#[derive(Debug, PartialEq, Default)]
+struct PravegaNodeUriParts {
+    scheme: Option<String>,
+    authority: Option<String>,
```

Comment:
Is this actually optional?

Reply:
These members are Optional because their corresponding components may not be present when `uri_parts_from_string` decodes the string.

Reply:
As we discussed, I think we can split this into two different objects, one internal to refer to segment store instances and one external to be included in config to refer to controller instances. 
For the external one, we should parse early and return errors or panic if it is malformed. For the internal one we can do as before and have a simple string wrapper without a scheme.

Reply:
for external URI, it appears all the existing code doesn't seem to expect an error when the URI itself is constructed.  Instead, it appears that only `build()`  could return an Err. When you say "parse early", do you mean literally when the URI is constructed, or when `build()` is called on the ClientConfigBuilder?


e.g. this example

```rust
    let config = ClientConfigBuilder::default()
        .controller_uri("localhost:9090")
        .build()
        .unwrap();
```

Would it be reasonable to wait until `build()` is called before parsing the value passed to `controller_uri`?
I don't believe `.controller_uri()` can return an Err, so it could only panic if you wanted to parse the URI at the time it is instantiated.

Reply:
Probably best to do it in the validation of the config. That is where an error is most expected.

Reply:
validation is done in `build()` step. 

I could not find where `segment store`  endpoints are returned from the controller, so I could not create a new/simpler, endpoint/uri type for that use case.  `ConnectionFactory` seems like where an endpoint would be used, but I could not figure out which call(s) to `establish_connection` would need to change to use the new type.